File: Python/12-Apple_and_Orange.py
# -*- coding: utf-8 -*-
""" 
Created on Tue Feb 21 2023

@author Carlos Páez
"""
import math

#
# Complete the `countApplesAndOranges` function below.
#
# The function accepts following parameters:
#   1. INTEGER s
#   2. INTEGER t
#   3. INTEGER a
#   4. INTEGER b
#   5. INTEGER_ARRAY apples
#   6. INTEGER_ARRAY oranges
#


def countApplesAndOranges(s: int, t: int, a: int, b: int, apples: list[int], oranges: list[int]):
    """
    It takes in a bunch of numbers, checks if they're valid, and then prints out the number of valid
    apples and oranges.

    :param s: starting point of Sam's house location
    :type s: int
    :param t: integer, the point where the house ends
    :type t: int
    :param a: starting point of Sam's house location
    :type a: int
    :param b: the location of the tree from which the oranges are thrown
    :type b: int
    :param apples: an array of integers representing the distances that each apple falls from point a
    :type apples: list[int]
    :param oranges: list[int]
    :type oranges: list[int]
    :return: The number of apples and oranges that fall on Sam's house.
    """
    if not all(1 <= param <= 10**5 for param in (s, t, a, b)):
        return 'Error'

    valid_apples = [
        apple for apple in apples
        if -10 ** 5 <= apple <= 10**5 and s <= apple + a <= t
    ]
    valid_oranges = [
        orange for orange in oranges
        if - 10**5 <= orange <= 10**5 and s <= orange + b <= t
    ]

    print(len(valid_apples))
    print(len(valid_oranges))


if __name__ == '__main__':
    first_multiple_input = input().rstrip().split()

    s = int(first_multiple_input[0])
    t = int(first_multiple_input[1])

    second_multiple_input = input().rstrip().split()

    a = int(second_multiple_input[0])
    b = int(second_multiple_input[1])

    third_multiple_input = input().rstrip().split()

    m = int(third_multiple_input[0])
    n = int(third_multiple_input[1])

    apples = list(map(int, input().rstrip().split()))
    oranges = list(map(int, input().rstrip().split()))

    countApplesAndOranges(s, t, a, b, apples, oranges)

File: JavaScript/12-Apple_and_Orange.js
/**
 * Created on Mon Feb 20 2023
 * 
 * @author Carlos Páez
 */
'use strict';

process.stdin.resume()
process.stdin.setEncoding('utf-8')

let inputString = ''
let currentLine = 0

/* Reading the input from the console. */
process.stdin.on('data', function (inputStdin) {
    inputString += inputStdin
})

/* Reading the input from the console. */
process.stdin.on('end', function () {
    inputString = inputString.split('\n')
    main()
})

/**
 * It reads the input string and returns the current line.
 * @returns The current line of the inputString.
 */
function readLine () {
    return inputString[ currentLine++ ]
}

/*
 * Complete the 'countApplesAndOranges' function below.
 *
 * The function accepts following parameters:
 *  1. INTEGER s
 *  2. INTEGER t
 *  3. INTEGER a
 *  4. INTEGER b
 *  5. INTEGER_ARRAY apples
 *  6. INTEGER_ARRAY oranges
 */

/**
 * It takes in a bunch of numbers, and returns the number of apples and oranges that fall on the house
 * @param s - starting point of Sam's house location.
 * @param t - integer, the number of steps that Sam's house is from the tree
 * @param a - starting point of house
 * @param b - the location of the tree on the x-axis
 * @param apples - [-2, 2, 1]
 * @param oranges - [-2, 2, 1]
 * @returns The number of apples and oranges that fall on Sam's house.
 */
function countApplesAndOranges (s, t, a, b, apples, oranges) {
    const minLimit = -1e5
    const maxLimit = 1e5

    for (const param of [ s, t, a, b ]) {
        if (!(1 <= param && param <= maxLimit)) return 'Error'
    }

    let validApples = 0
    let validOranges = 0

    for (const apple of apples) {
        if (minLimit <= apple && apple <= maxLimit) {
            if (s <= (apple + a) && (apple + a) <= t) validApples += 1
        }
    }

    for (const orange of oranges) {
        if (minLimit <= orange && orange <= maxLimit) {
            if (s <= (orange + b) && (orange + b) <= t) validOranges += 1
        }
    }

    console.log(`${ validApples }\n${ validOranges }`)
}


function main () {
    const firstMultipleInput = readLine().replace(/\s+$/g, '').split(' ')

    const s = parseInt(firstMultipleInput[ 0 ], 10)
    const t = parseInt(firstMultipleInput[ 1 ], 10)

    const secondMultipleInput = readLine().replace(/\s+$/g, '').split(' ')

    const a = parseInt(secondMultipleInput[ 0 ], 10)
    const b = parseInt(secondMultipleInput[ 1 ], 10)

    const thirdMultipleInput = readLine().replace(/\s+$/g, '').split(' ')

    const m = parseInt(thirdMultipleInput[ 0 ], 10)
    const n = parseInt(thirdMultipleInput[ 1 ], 10)

    const apples = readLine().replace(/\s+$/g, '').split(' ').map(applesTemp => parseInt(applesTemp, 10))
    const oranges = readLine().replace(/\s+$/g, '').split(' ').map(orangesTemp => parseInt(orangesTemp, 10))

    countApplesAndOranges(s, t, a, b, apples, oranges)
}